GEP-1762: In Cluster Gateway Deployments

howardjohn opened the pull request:
Review note: This GEP was split into 3 - the current GEP (focusing on in-cluster deployments), https://github.com/kubernetes-sigs/gateway-api/pull/1868 (infrastructure field we depend on), and https://github.com/kubernetes-sigs/gateway-api/pull/1863/files (gateway merging)

**What type of PR is this?**

/kind gep

**What this PR does / why we need it**:

This PR attempts to inject some opinions about how in-cluster deployments should work. Currently, there is a lot of different implementations behaving differently, leading to confusion and drifting user experiences.

As-is, this PR is fairly opinionated. I expect that some of the MUSTs become SHOULDs as we iterate on this PR. In its current state, this is largely meant as a discussion point. I expect there to be a lot of iteration as we learn different implementations requirements and perspectives.

Note: most of the names were just picked arbitrarily. I'd like to first focus on the concepts, then we can debate the best names (almost certainly not what the initial GEP has!)

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, please enter a release note below:
-->
```release-note

```

robscott commented on site-src/geps/gep-1757.md:
```diff
@@ -0,0 +1,224 @@
+### Resource Attachment
+
+Generated resources are guaranteed to have two attributes:
+* A `gateway.networking.k8s.io/metadata.name: <NAME>` label.
```
To clarify, this is only for resources generated for a Gateway? If there exists a future use case where resources are generated for some other resource in Gateway API, would we have a way of further distinguishing that? Also have a similar question about if we should try to specify the namespace of the original resource or if that's always going to be the same.

howardjohn replied:
Yes, this is for resources generated for the Gateway. IE Service and Deployment.

I was assuming same namespace. I can put that in the spec, although it may be too limiting. So far I am being opinionated until someone suggests otherwise though, so I will make it a MUST for now...

robscott commented on site-src/geps/gep-1757.md:
```diff
@@ -0,0 +1,224 @@
+
+This document uses a few terms throughout. To ensure consistency, they are defined below:
+
+* In-cluster deployment: refers to an implementation that actuates a `Gateway` by running a data plane in the cluster.
+  This is *often*, but not necessarily, by deploying a `Deployment`/`DaemonSet` and `Service`.
+* Automated deployment: refers to an implementation that automatically deploys the data plane based on a `Gateway`.
+  That is, the user simply creates a `Gateway` resource and the rest is handled behind the scenes by the implementation.
+* Manual deployment: refers to an implementation that does NOT automatically deploy the data plane based on a `Gateway`.
+  This may require, for example, a user to manually create a `Deployment` and `Service` for the data plane, then link this to a `Gateway`
```
Nit: Maybe using `In-cluster Dataplane` or something similar would help differentiate that the first item in this list is actually a different dimension than the following two. (IE you're not choosing one of In-Cluster, Automated, or Manual).

robscott commented on site-src/geps/gep-1757.md:
```diff
@@ -0,0 +1,224 @@
+
+## Overview
+
+The gateway-api provides a common abstraction over different implementations, whether they are implemented by cloud load balancers, in-cluster deployments, or other mechanisms. However, many in-cluster implementations have solved some of the same problems in different ways.
```
Tiny nit: Prefer to use "Gateway API" over variations with "-" or "The".

robscott commented on site-src/geps/gep-1757.md:
```diff
@@ -0,0 +1,224 @@
+When an implementation uses `Service`:
+* MUST have `Spec.Address` be a `NamedAddress` referring to the Service by name. Cross namespace references are not allowed.
+
+Any other "existing infrastructure" types re implementation specific
```
```suggestion
Any other "existing infrastructure" types are implementation specific.
```

robscott commented on site-src/geps/gep-1757.md:
```diff
@@ -0,0 +1,224 @@
+
+With any in-cluster deployment, customization requirements will arise. 
+
+Some common requirements would be:
+* `Service.spec.type`, to control whether a service is a `ClusterIP` or `LoadBalancer`.
+* If it is a `LoadBalancer`, the `Service.spec.loadBalancerIP` to assign to it.
+* Arbitrary labels and annotations on generated resources.
+* CPU and memory requests, and other similar customizations.
```
Would providing a standard way to configure these concepts be in scope for this GEP?

howardjohn replied:
Yes absolutely. But I don't have any great ideas yet :slightly_smiling_face:

robscott commented on site-src/geps/gep-1757.md:
```diff
@@ -0,0 +1,224 @@
+* Arbitrary labels and annotations on generated resources.
+* CPU and memory requests, and other similar customizations.
+
+For setting a specific IP address in the Service, the `.Spec.Address` field can be used. For example:
```
To clarify, is this trying to attach a Gateway to an existing Service with the specified ClusterIP?

howardjohn replied:
I was a bit worried this was confusing... might need refinement

NamedAddress = manual deployment, reference existing Service

IPAddress = spin up a new Service, but set the loadBalancerIP to this specific IP address already.

robscott commented on site-src/geps/gep-1757.md:
```diff
@@ -0,0 +1,224 @@
+This differs from the selection of "manual deployment", which doesn't use `IPAddress` type.
+
+Currently, all other customizations are implementation specific.
+Temporary Note: it seems like most implementations are using `parametersRef` in GatewayClass. The author thinks this is a bit problematic.
```
I think parametersRef can be a useful mechanism, but would be helpful to get feedback from implementations using it today. I do agree with the overall fear of introducing too many GatewayClasses or requiring too many people to have cluster-scoped access to work with this API, but I think we can find some kind of balance here that doesn't completely exclude GatewayClass.

robscott commented on site-src/geps/gep-1757.md:
```diff
@@ -0,0 +1,224 @@
+
+Generated resources are guaranteed to have two attributes:
+* A `gateway.networking.k8s.io/metadata.name: <NAME>` label.
+* A name `<NAME>-<GATEWAY CLASS>`.
```
If we were to use this pattern we'd need to define some kind of patterns for when we hit the max length of a resource name.

robscott commented on site-src/geps/gep-1757.md:
```diff
@@ -0,0 +1,224 @@
+## API
+
+This GEP does not introduce any API changes to the spec. However, it does add new requirements for use of the same schema.
+These requirements will be codified as conformance tests.
```
How would we write conformance tests for this?

robscott commented on site-src/geps/gep-1757.md:
```diff
@@ -0,0 +1,224 @@
+# GEP-1757: In Cluster Gateway Deployments
```
Nit: GEPs should be tied to an issue number so we can track the lifecycle of a GEP after the initial PR is merged.

howardjohn replied:
Ah that makes much more sense than making a dummy PR to get a number assigned :upside_down_face: thanks, will change it

robscott commented on site-src/geps/gep-1757.md:
```diff
@@ -0,0 +1,224 @@
+  name: my-gateway
+spec:
+  addresses:
+  - type: NamedAddress
+    value: some-existing-service
```
This was definitely not the original intent of NamedAddress, we may want to use a different address type if this is a workflow we want to support.

howardjohn replied:
I realized after its technically deprecated as well, so I would agree a new Address type is probably best (or equivilent).

robscott commented on site-src/geps/gep-1757.md:
```diff
@@ -0,0 +1,224 @@
+
+### Merging Gateways
+
+The spec currently specifies that gateways *can* be merged. This MUST be implemented by referencing the same "existing infrastructure" (as defined in "Manual Deployments").
```
To clarify, you're saying that this must be implemented in a certain way only _if it is supported_?

howardjohn replied:
Yep should add that; thanks

robscott commented on site-src/geps/gep-1757.md:
```diff
@@ -0,0 +1,224 @@
+  name: merged-gateway
+spec:
+  addresses:
+  - type: NamedAddress
+    value: parent-gateway
```
Above you're using NamedAddress to reference a Service name and now it's a Gateway name, would these ideas be compatible?

robscott commented on site-src/geps/gep-1757.md:
````diff
@@ -0,0 +1,224 @@
+```
+
+With this configuration, an implementation:
+* MUST treat the `parent-gateway` as having the concatenated list of all listeners from merged Gateways.
+  This means, for example, that a Service with port 80 and 81 may be created in this example.
````
Conflict resolution seems like it would be pretty tricky here. The idea of merging Gateways probably could require an entirely separate GEP.

howardjohn replied:
I agree. It is already in the spec though (but under specified). I would be fine with ripping this section out entirely possibly

dprotaso replied:
Prior issue: https://github.com/kubernetes-sigs/gateway-api/issues/1713

robscott commented on site-src/geps/gep-1757.md:
```diff
@@ -0,0 +1,224 @@
+spec:
+  # Match the generated Deployment by reference
+  # Note: Do not use `kind: Gateway`.
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: gateway-example
```
Maybe we could introduce a simple mutating webhook that would translate Gateway references to this format?

howardjohn replied:
Maybe. Could be a bit confusing though when user looks at it (also its stateful since you need to lookup the gateway class name).

IMO making Gateway implement /scale would be pretty neat

robscott reviewed (commented):
Thanks for the work on this @howardjohn!

howardjohn commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+  This MUST be derived from the referenced `Spec.Address`.
+* MUST not deploy any resources into the cluster; it is expected that a user will do these actions.
+
+### Merging Gateways
```
Note: this probably needs to be broken out, it applies to all implementation types. However, given the `infrastructure` field is shared with the rest of the GEP. I intent to keep it in here for the initial discussions to keep conversation focused

youngnick replied:
Yeah, it seems like this would be right at home in a GEP about `infrastructure` by itself, as another use case to justify including the new struct.

sunjayBhatia commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+  name: merged-gateway
+spec:
+  infrastructure:
+    attachTo:
```
this primary Gateway idea is interesting, makes some of the precedence more explicit when merging which is great (relying on the the general conflict resolution guidelines where the resource that comes "first" will still apply in some cases it seems)

costinm replied:
I'm a bit concerned with this as a requirement - allowing implementations freedom to merge as needed is important.

A multi-tenant, external gateway implementation may combine all the gateways of type http with a valid domain name. Even a single-in-cluster gateway could do the same. Note that implementations are not bound by namespace or even cluster or project/owner boundaries. 

There is a separate problem of 'too many resource to fit in a single yaml' - but creating 2 Gateways will result in 2 IPs and deployments. A simple naming pattern ( same prefix ) or common label may work too, infrastructure.attachTo seems a bit heavy.

costinm replied:
Or even better - if you add 'attachTo. Service' in the section above, having 2 Gateways attach to the same Service will be a clear signal that they need to be merged ( or an error condition if they must explicitly attach to Gateway ).

youngnick replied:
Yes, the idea behind `attachTo` is to allow multiple Gateways to attach to something that provides the address information, which could be another Gateway or a Service (presumably that would be manually configured). I have some concerns, but I think I just need to think through the use case, and on balance think it's pretty reasonable.

sunjayBhatia commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+With any in-cluster deployment, customization requirements will arise. 
+
+Some common requirements would be:
+* `Service.spec.type`, to control whether a service is a `ClusterIP` or `LoadBalancer`.
```
in Contour we've made this configurable via GatewayClass params but as you say in other places it is a bit unwieldy, esp. as you add in the desire to configure a particular Service port and NodePort which should be done at the Gateway level

youngnick replied:
I'd expect that in the case we add `infrastructure`, that any existing GatewayClass `paramsRef` config that overlaps would be treated as an implementation-wide _default_, that could be _overridden_ by the more-specific Gateway setting.

youngnick commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+      kind: Service
+      name: existing-service
+  addresses:
+  - type: Service
```
This implies adding a `type` of `Service`, which I'm not opposed to necessarily, but it seems like syntactic sugar for `ClusterLocalIPAddress`, basically?

howardjohn replied:
I did that at first but realized its definitely not the same. Its far more common for the Service referenced to be a LoadBalancer type

robscott replied:
Why do we need both this and the `attachTo` mechanism?

youngnick replied:
Yeah, I think as @robscott alludes to in another comment, it would be better to support `attachTo` in `infrastructure`, and add additional support to the `address` type for more types of address that aren't about attachment.

howardjohn replied:
I think I simply forgot to remove `address`  when I changed it to infrastructure

costinm commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+With this configuration, an implementation:
+* MUST mark the Gateway as Ready and provide an address in `Status.Addresses` where the Gateway can be reached on each configured port.
+* MUST label all generated resources (Service, Deployment, etc) with `gateway.networking.k8s.io/metadata.name: my-gateway` (where `my-gateway` is the name of the Gateway resource).
+* MUST name all generated resources `my-gateway-example` (`<NAME>-<GATEWAY CLASS>`).
```
Or use that as a prefix ( my-generated-example-v2 for a deployment when doing a rollout ).

danehans replied:
+1 on prefix of `<GATEWAY_NAME>-<GATEWAYCLASS_NAME>-` or should the name be least -> most specific by prefacing the resource name with `<GATEWAYCLASS_NAME>-<GATEWAY_NAME>-`?

danehans replied:
Generated names can get quite lengthy. Should generated names that are greater than a defined length get hashed? If so, include these details in the spec.

howardjohn commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+
+With this configuration, an implementation:
+* MUST mark the Gateway as Ready and provide an address in `Status.Addresses` where the Gateway can be reached on each configured port.
+* MUST label all generated resources (Service, Deployment, etc) with `gateway.networking.k8s.io/metadata.name: my-gateway` (where `my-gateway` is the name of the Gateway resource).
```
there is ownerRef but, IMO, its useful to have a label since a lot of things need label selector (HPA, for example). Plus deployment itself needs a label selector to match on Pods, as well as Service

costinm replied:
But we'll still have ownerRef too ? 

I don't mind having a common label - it would also work for the other case in this doc ( merging gateways, since the 'other' gateways are also associated with the same gateway ).

costinm replied:
We can't cross namespace boundary easily ( i.e. a setting in ns1 shouldn't be able to influence something in ns2 ).

We certainly need to support gateways running outside of the namespace (or cluster) - but with care.

youngnick replied:
Yes, I think that we should not allow for namespace-crossing with the deployment without a very strong use case. As we've seen from the issues that led to ReferenceGrant, allowing cross-namespace references can have large unintended consequences. I think we should leave this as "same namespace as Gateway" unless there's a very strong use case otherwise.

danehans replied:
I'm not opposed to restricting to the same ns as a Gateway. My thinking is to start out permissive, see if any use cases arise, and then become more restrictive.

youngnick replied:
One lesson I've learned doing this for a while now, is it's easier to start restrictive and carve out exceptions if they're necessary - the opposite is suprisingly difficult! That's why I'd prefer to rule out namespace-crossing without a strong use case - we can carve out exceptions later.

hzxuzhonghu replied:
+1 to have ownerref as well, it is easy to be gced when gateway is deleted

costinm commented on site-src/geps/gep-1762.md:
````diff
@@ -0,0 +1,309 @@
+```
+
+With this configuration, an implementation:
+* MUST mark the Gateway as Ready and provide an address in `Status.Addresses` where the Gateway can be reached.
````
Can the user manually creating the Gateway just populate spec.Address or status.Address ? Would work for external LBs too.  

But I'm not sure we should be prescriptive about 'self deployed' or 'external' - better to focus on the 'auto-deployed'. There are many other valid options - could be a DNS name for example.

howardjohn replied:
Yes, that is the idea - to allow users to set it and align with external LBs. In both case you just set spec.Address.

(.Status.Address is set to the assigned address, .Spec.Address is user intent)

costinm replied:
Yes, but if the user somehow manages the assignment - they would set Status.Address too ( gateway controller may not even have a way to find it - for example an external separate 'front' DNS and LB )

youngnick replied:
If the user is prepared to manage the complexity of using the `status` subresource to update the `.status.address` fields, then sure. But I think that will be uncommon.

costinm replied:
"User" is likely to mean some tool or CI/CD - or other controllers different from the gateway controller.
For example based on namespace and permissions something may provision a DNS name, get ACME certs, setup some infra - and populate the status.address.  

 Unfortunately K8S is very backwards in using IP addresses directly - using domain names with shared IPs for http has been the practice for many years now, I don't know any other modern system still having users deal with IP addresses.

danehans replied:
The address can be a [domain name](https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1alpha2/shared_types.go#L362). IMO most implementations want to manage status and not defer certain fields to another controller/tool. For the use case above, one option is for the tool to create a custom resource that represents the external addresses/names. This custom resource would be a local object reference of `gateway.spec.addresses[]`. When the status of this custom resource is `Ready=true`, the Gateway controller sets `status.addresses`.

youngnick replied:
In my comment, I was talking about the case where a user _manually_ creates a Service object for some reason (which would be pretty weird imo).

I think in the case of Gateway, the reason to use IP addresses is to allow for other things (like external-dns) to handle the name management for you. Because Gateway deals in vhost names, _not_ having an IP address would be pretty weird here too.

Both Gateway API implementations that I've worked on, that use Loadbalancer Services for the actual data plane traffic path creation, have just picked up the Loadbalancer Service details, lightly translated them, and then set the status manually. Which means that on AWS, if you use an ELB or ALB, you'll get a hostname (as @danehans reminds us) instead of an IP.

costinm commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+This introduces a new concept in Gateway merging of a "primary" Gateway.
+This makes it explicit which is the primary gateway and which are attached to it.
+
+Secondary Gateways cannot configure `addresses` or anything in `infrastructure` (aside from `attachTo`), and must have
```
I assume secondary gateways will have the status populated from the parent ? Users should not have to follow the links.

howardjohn replied:
Yeah I think that makes sense, I will update this to be more precise about how status works

costinm commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+spec:
+  addresses:
+  - type: IPAddress
+    value: 1.1.1.1
```
What happens if 2 Gateways spec the same IPAddress ? Is this another signal that they need to be merged ? I don't think it can be an error condition - for externally managed / multi-tenant it's the most common case.

costinm commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+
+#### Arbitrary Customization
+
+Currently, to provide arbitrary customizations, Gateway API provides a few mechanisms:
```
Not sure if 'arbitrary customization' should be in scope of the Gateway API. 

Since we support 'external' gateways - each implementation may have its own APIs and configs - in many cases outside of K8S. For in-cluster - if we allow users to create their own deployment and Service - they already have all they need to do arbitrary customization. 

I don't see a middle case where more arbitrary customizations would be needed - nor the use case for gateway API to prescribe how implementations handle their custom configs.

howardjohn replied:
Example use case: I want to set CPU requests to "2 CPUs". How can I do that?

If I create a deployment before I deploy the Gateway, I don't know what to put in there - there are 100s of fields like image, etc that I want the controller to apply for me.

If I patch the deployment afterwards, this has a few issues:
* Controller cannot do sophisticated merging logic. It cannot say "I want to default 1 CPU unless user has set it"; it can either force ownership or never set it, afaik
* We have some period of time with the wrong settings. For CPu its not so bad (besides needing to redeploy for no reason), but what if the user's customization was "do not run as root" for example

Both approaches are also tricky since they rely on ordering between actions, which is not very gitops/declarative friendly.

costinm replied:
I understand - my point was that it may be a broader problem ( anything that creates resources has it ), and may be handled by each implementation. 

For example in Istio we had a helm chart to install gateway without injection or controler involvment. Others may still do this, and it takes care of all options including image.

Having some template is also common - Deployment, Knative, etc have spec with a lot of options. Just not sure if we want to mix this broad problem into this specification. Would be a good thing to solve, but in a separate context ( and as for other proposals - after some survey on how different vendors deploy in cluster gateways ).

costinm commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+* A `gateway.networking.k8s.io/metadata.name: <NAME>` label.
+* A name `<NAME>-<GATEWAY CLASS>`.
+
+The generated resources MUST be in the same namespaces as the `Gateway`.
```
We should probably discuss this, may be too expensive and not clear if it is necessary. 

If we allow external multi-tenant - we can probably support per-cluster gateway too. Implementations should be free to generate the deployment in a central namespace if they can preserve the isolation semantics.

youngnick replied:
I'm very reluctant to allow ways to _explicitly_ request crossing namespace boundaries in the absence of a two-way handshake (which is not practical here). Maybe a better way to write this is "_If_ Kubernetes resources are generated to satisfy a Gateway request, they MUST be in the same namespace as the Gateway".

Cilium's Gateway API support doesn't generate a Deployment to provision an Envoy instance, but does provision CiliumEnvoyConfig resources to control the per-node Envoy instances, and _those_ are created in the same namespace as the Gateway.

danehans commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+* MUST name all generated resources `my-gateway-example` (`<NAME>-<GATEWAY CLASS>`).
+  This is not simply `NAME` to reduce the chance of conflicts with existing resources.
+
+### Manual Deployments
```
It might be helpful to link https://github.com/kubernetes-sigs/gateway-api/issues/1687 regarding manual deployment since this concept is currently unspecified by the API.

robscott commented on site-src/geps/gep-1762.md:
````diff
@@ -0,0 +1,309 @@
+```
+
+With this configuration, an implementation:
+* MUST mark the Gateway as Ready and provide an address in `Status.Addresses` where the Gateway can be reached on each configured port.
````
```suggestion
* MUST mark the Gateway as Programmed and provide an address in `Status.Addresses` where the Gateway can be reached on each configured port.
```

robscott commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+metadata:
+  name: my-gateway
+spec:
+  infrastructure:
+    attachTo:
```
I think the `infrastructure` addition to Gateway really is the core of this GEP so it would be helpful to spell out the Go types associated with it. If not in this PR, at least as an AI before the GEP can transition from "provisional" to "experimental".

danehans replied:
@robscott thanks for the clarification. IMHO it makes sense to spin off a separate KEP that defines the manual deployment model. @howardjohn let me know if I can help with any of the spin-off KEPs of this PR.

danehans replied:
In this example, will the Gateway will use all addresses from the Service? I see `listeners[].port` is still being defined since it's required. Why not use the Service port too?

howardjohn replied:
Added some references to the other GEPs. Fortunately those have progressed a lot since this comment

robscott commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+
+With any in-cluster deployment, customization requirements will arise. 
+
+Some common requirements would be:
```
To clarify, if we can attach a Gateway to an existing Service, do we need to allow anything else to be configurable on the Gateway, or should we just rely on users provisioning custom Services whenever customization is needed?

costinm replied:
Good point - a Service of type LoadBalancer with selector and associated deployment ( created by user ) and attached Gateway is a very reasonable model. Gateway can customize the TLS server setting and define routing, but can reuse the Service address and deployment.

howardjohn replied:
Yes, we do. There is a big difference between "Manual deployment" and "Customized deployment".

Lets say I want to set `sessionAffinity` on service. First, I need to go create my Service. I guess I know port 443 will be used in the gateway, so I create the Service and link it up to the Gateway.

A week later, I add port 80 to the Gateway as well. Now its broken; I need to keep the Service in sync.

With _customization_, a user would simply configure it wants sessionAffinity, but the implementation would be responsible for keeping everything else in sync.

robscott commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+
+Some common requirements would be:
+* `Service.spec.type`, to control whether a service is a `ClusterIP` or `LoadBalancer`.
+* IP in the Service to assign to it.
```
If we had a way to represent it, it seems like the simplest answer may just be to have a way to represent all kinds of addresses that could be represented on a Service, and then attach all of them to the Gateway as well.

youngnick replied:
Keeping in mind that "no address in spec" means "I don't care what the address is", for the record.

robscott commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+
+#### Labels and Annotations
+
+Labels and annotations SHOULD be copied from the `Gateway` to generated resources.
```
This feels like it may lead to some undesired behaviors. Although I can see the rationale for this in some cases, I'd hate to include the suggestion in the spec unless we were very confident about it.

youngnick replied:
Yeah, agreed. It's fine for implementations to choose to do it, but I wouldn't want it to be any stronger than a _may_.

robscott commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+  This *can* be used to make arbitrary customizations, but comes with a number of issues:
+  * GatewayClass is cluster scoped. This means that any user configuring a namespace-scoped Gateway
+    may need to make a new cluster-scoped resources, which makes RBAC extremely challenging.
+    An analogy would be to have a `DeploymentClass` to configure CPU requests, instead of a field in each `Deployment`.
```
Well this is fair, we do have many workload resources that are used to provide different ways to schedule Pods (Deployment, StatefulSet, DaemonSet, etc). With the resource class model, it would be reasonable to say that these concepts would map to different Gateway Classes. Of course my reference example here is Internal vs External LBs which often represent different kinds of underlying infra. So while I agree that we probably don't want to represent something like Resource Requests with a separate GWC, I don't think we want to discount the use of GWC entirely.

robscott commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+  * GatewayClass is cluster scoped. This means that any user configuring a namespace-scoped Gateway
+    may need to make a new cluster-scoped resources, which makes RBAC extremely challenging.
+    An analogy would be to have a `DeploymentClass` to configure CPU requests, instead of a field in each `Deployment`.
+  * Many levels of indirection. To configure my `Gateway`, I need a `Gateway`, `GatewayClass`, and `SomeCustomParamsType`.
```
On the other hand, it's much easier for implementations to bundle a set of commonly used configurations with different preinstalled GatewayClasses. This would be much harder to represent with the per-Gateway config proposed with this GEP.

youngnick replied:
I said this elsewhere, but I think the expected behavior should be that more specific beats less specific, meaning that `infrastructure` fields would override GatewayClass parameters to some extent. Obviously, this will depend on the implementation though (maybe you don't want your `internal` GatewayClass provisioning Loadbalancer Services...)

howardjohn replied:
+1, I will explicitly call this out

robscott commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+    may need to make a new cluster-scoped resources, which makes RBAC extremely challenging.
+    An analogy would be to have a `DeploymentClass` to configure CPU requests, instead of a field in each `Deployment`.
+  * Many levels of indirection. To configure my `Gateway`, I need a `Gateway`, `GatewayClass`, and `SomeCustomParamsType`.
+  * This is intended to be immutable according to the `GatewayClass` spec. This makes it challenging to operate and implement.
```
I think the goal is for GatewayClass + Params to be considered a point-in-time template for Gateway creation. If we have any requirements for immutability here I think that likely needs to be changed.

robscott commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+    An analogy would be to have a `DeploymentClass` to configure CPU requests, instead of a field in each `Deployment`.
+  * Many levels of indirection. To configure my `Gateway`, I need a `Gateway`, `GatewayClass`, and `SomeCustomParamsType`.
+  * This is intended to be immutable according to the `GatewayClass` spec. This makes it challenging to operate and implement.
+  * There is no common type to reference; each implementation would make their own.
```
Although there aren't common top level fields on GatewayClass today, we can always add some if they fit.

robscott commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+  * Order may matter; if implementation is first we may temporarily have unsafe settings (for example, run as root later overwritten to run as non-root).
+  * This type of ordering dependency is tricky operationally, especially with GitOps.
+
+None of these mechanisms provide a satisfying solution.
```
A lot of the content in the section above feels like it belongs in an `Alternatives Considered` section.

robscott commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+      # Define the version to use. The meanings of version is implementation specific
+      version: v1.2
+      # Number of replicas for the Gateway. The meaning is implementation specific; for Deployments this SHOULD be `.spec.replicas`.
+      replicas: 2
```
If we were to allow this to be configurable, maybe something more generic like `size` would more easily map to non-deployment use cases.

robscott commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+  spec:
+    infrastructure:
+      # Define the scope of the Gateway (GEP-1651)
+      scope: InCluster
```
I think I'd personally rather see this as per-address config or a GatewayClass field.

robscott commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,309 @@
+as some implementations may have different cleanup mechanisms.
+
+
+These can be relied on to attach resources to.
```
This sentence feels a bit out of place, not sure what "these" refers to here.

robscott reviewed (commented):
Thanks @howardjohn! Lots of interesting ideas here. Will probably need to split this up eventually, but helpful to see everything together for now.

youngnick commented:
I took another pass, I think that between this PR and #1653, we actually have solutions for #1762, #1651, and #1713.

For next steps, I think that we should break out the `infrastructure` stanza into its own GEP, get that agreed and in, and then round out #1713. Between `infrastructure` and having address scopes (and maybe some more types), I think that we can cover everything in this GEP in a much smaller document, basically just specifying how those two features would work together to achieve this use case.

danehans commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,312 @@
+With this configuration, an implementation:
+* MUST mark the Gateway as `Programmed` and provide an address in `Status.Addresses` where the Gateway can be reached on each configured port.
+* MUST label all generated resources (Service, Deployment, etc) with `gateway.networking.k8s.io/metadata.name: my-gateway` (where `my-gateway` is the name of the Gateway resource).
+* MUST provision generated resources in the same namespace as the Gateway.
```
What about cluster-scoped resources?

danehans commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,312 @@
+
+### Manual Deployments
+
+When a manual deployment is used, a user must explicitly link their Gateway to existing infrastructure. This is done by specifying `Spec.Address`.
```
The example below uses `spec.infrastructure.attachTo` and not `spec.address`. Can you clarify the proposed schema?

shaneutt commented on site-src/geps/gep-1762.md:
```diff
@@ -0,0 +1,312 @@
+* [Support cluster-local Gateways](https://github.com/kubernetes-sigs/gateway-api/discussions/1247)
+* [Scaling Gateway Resources](https://github.com/kubernetes-sigs/gateway-api/discussions/1355)
+* [Manual deployments](https://github.com/kubernetes-sigs/gateway-api/issues/1687)
+
```
It would appear since this GEP is getting into the `automated` vs. `manual` provisioning of `Gateway` discussions that we should reference  #1687 which was asking for definitions and documentation in this regard (and this GEP would appear to be the first step in supporting that):

```suggestion
* [Static/Unmanaged Gateway Mode](https://github.com/kubernetes-sigs/gateway-api/issues/1687)
```

shaneutt reviewed (approved):
The overview, goals and terminology all make sense to me. I think for this initial PR we should have probably stopped at those because as we can see the deeper discussion we added right away immediately slowed down the velocity with long comment conversations that I believe would have been better in targeted sessions.

I have one suggestion about adding a link I would like to see added if you don't mind adding that suggestion. However, given that this is `Provisional` and I what I would like to see is smaller PR follow ups which continue to tweak and tease out the details until we're comfortable with calling it `implementable`, I am approving this. There are some details of what was written that I too would like to maybe see changed/tweaked, but I am comfortable with adding my own follow-up PRs to propose them.

I'm really glad to see that we're adding this GEP to get these things better defined, thanks for starting this!

dprotaso commented:
I pulled the merging Gateways into it's own GEP doc - https://github.com/kubernetes-sigs/gateway-api/pull/1863

howardjohn commented:
Review note: This GEP was split into 3 - the current GEP (focusing on in-cluster deployments), https://github.com/kubernetes-sigs/gateway-api/pull/1868 (infrastructure field we depend on), and https://github.com/kubernetes-sigs/gateway-api/pull/1863/files (gateway merging)

This GEP is now much more narrowly scoped.

pleshakov commented on site-src/geps/gep-1762.md:
````diff
@@ -0,0 +1,228 @@
+```
+
+With this configuration, an implementation:
+* MUST mark the Gateway as Ready and provide an address in `Status.Addresses` where the Gateway can be reached.
````
> * MUST mark the Gateway as Ready 

Even though the infrastructure is running in the cluster (no need to provision it), I don't think all implementations can say it is READY. For example, for implementations that rely on eventually consistent propagation of data plane configuration to multiple data plane replicas, it is still hard to tell if those data planes have been configured successfully and ready to handle the traffic.

howardjohn replied:
The meaning of Ready changed since I wrote this. I have updated this

howardjohn commented:
hi folks, what is needed to push this forward? Thanks

youngnick commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,228 @@
+
+#### Labels and Annotations
+
+Labels and annotations SHOULD be copied from the `Gateway` to generated resources.
```
I'm a -1 on this, I'd rather see us add standard `labels` and `annotations` to `infrastructure` that would then be copied to generated resources. That puts the desired labels and annotations into the `spec` of the Gateway, which makes sense because they are a user request. They can still be a `map[string]string` though, I think that's one of the only cases that we should use them.

howardjohn replied:
That seems reasonable to me. will update

howardjohn replied:
done

youngnick commented:
Aside from copy annotations from the Gateway, this LGTM.

howardjohn commented:
I have updated this to take advantage of the now-merged `infrastructure` field and added some clarity to the API spec as a result. No major changes though, just more description.

I don't think there are pending comments, so ptal when you have a chance @youngnick @robscott @shaneutt

rainest commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,281 @@
+* MUST label all generated resources (Service, Deployment, etc) with `gateway.networking.k8s.io/metadata.name: my-gateway` (where `my-gateway` is the name of the Gateway resource).
+* MUST provision generated resources in the same namespace as the Gateway if they are namespace scoped resources.
+  * Cluster scoped resources are not recommended.
+* MUST name all generated resources `my-gateway-example` (`<NAME>-<GATEWAY CLASS>`).
```
How should this handle multiple resources of the same kind without colliding? Was this intended to be something like "MUST use <pattern> as a name prefix for generated resources"?

We, for example, automatically create multiple Deployments (the proxy instances and a configuration manager) and and multiple Services (admin traffic, proxy traffic, and some utility/status APIs), some associated with one Deployment and some with the others.

We currently do something like the prefix approach. IIRC it's caused some issues because of the name character limit, and I don't think we've found a better way around that than telling users to choose shorter names.

howardjohn replied:
Currently it does not. Interesting case. On benefit of deterministic names is a lot of other policies attach via name. For example HPA. So having a common name is pretty handy. That is tricky if we have multiple though...

shaneutt replied:
Is this dilemma something we can capture in a TODO section so that we can merge, but don't have to hold up this specific PR?

shaneutt replied:
I remain in favor of creating a TODO item for this in the GEP, with the intention that we don't move to `implementable` until we resolve this, but to allow us to move forward with this first large iteration and resolve this separately as its own focus.

robscott replied:
I think it's going to be very difficult to _require_ this behavior without being disruptive to all in-cluster implementations. I think suggesting a pattern, and stating that it may just serve as a prefix (ie `generateName` instead of `name`) may be sufficient here.

rainest commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,281 @@
+* MUST provide an address in `Status.Addresses` where the Gateway can be reached.
+  This MUST be derived from the referenced `spec.infrastructure.attachTo`.
+* MUST not deploy any resources into the cluster; it is expected that a user will do these actions.
+* MUST reject any other `infrastructure` fields being configured; these will not be respected.
```
This impact of this requirement isn't clear if you're not already familiar with the other fields, so for reference:

```suggestion
* MUST reject any other `infrastructure` fields (see [GEP-1867](https://gateway-api.sigs.k8s.io/geps/gep-1867/#api)being configured; these will not be respected.
```

rainest commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,281 @@
+With this configuration, an implementation:
+* MUST provide an address in `Status.Addresses` where the Gateway can be reached.
+  This MUST be derived from the referenced `spec.infrastructure.attachTo`.
+* MUST not deploy any resources into the cluster; it is expected that a user will do these actions.
```
Are there any cases where it would not create, but would modify those resources? I think the answer is no, unless it's metadata only (label edits or similar).

howardjohn replied:
I switched to "must not create or modify". I don't mind allowing metadata as well but we can probably keep it small to start

rainest commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,281 @@
+  name: my-gateway
+spec:
+  infrastructure:
+    attachTo:
+      kind: Service
+      name: existing-service
```
This is currently a single object, but adequately referencing resources may require multiple.

In Kong's case, we validated that the resources associated with our manually-deployed Gateway could satisfy the Gateway (i.e. that the indicated Listeners were actually configured). We needed to check both the Service and the Deployment, since the former handles the LB provisioning and the latter handled application-level port binds.

howardjohn replied:
Yeah I guess we could extend it. Although I am curious why you need to reference the Deployment for ports - isn't that handled by TargetPort?

rainest replied:
It's for the case where you've done something like `targetPort: 80` in the Kubernetes configuration, but don't have something equivalent to a `listen 80` in an nginx.conf. We want to ensure there's a matching configuration for the target ports.

howardjohn replied:
Ah in our case the control plane configures the proxy config, so its impossible to end up in that state. Thanks for the info

rainest reviewed (commented):
Some overlap with prior comments since I didn't think to look for all the comments that Github hides after they become outdated, but hopefully useful thoughts based on our own experience building a manually-provisioned instance for most of our implementation history.

gcs278 commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,281 @@
+
+#### Gateway IP
+
+For setting a specific IP address in a generated `Service`, the `.Spec.Address` field can be used.
```
just a *nit*
```suggestion
For setting a specific IP address in a generated `Service`, the `.Spec.Addresses` field can be used.
```

shaneutt commented:
I personally think since this is `Provisional` and the PR has been open for a very long time, it should be reasonable to merge in its current state and we allow some time for it to soak.

/approve
/lgtm

However I'll hold to make sure either @robscott or @youngnick or both agrees.

/hold

shaneutt commented:
@howardjohn it looks like there are a couple of loose comments that need resolving, and then LGTM as a first provisional merge.

/approve
/lgtm

Hold for comments to be resolved, and for either @robscott or @youngnick (or both) to approve as well.

/hold

howardjohn commented:
Came back to this and gave it another pass. I believe the comments are either resolved or are referring to outdated info. I struggled to actually see all the comments due to GitHub's excellent UI, though

shaneutt reviewed (approved):
I'm not huge on having API details in the first PR.

That said, given that this is `Provisional` and has been hanging around a LONG time I'm in favor of merging it so that we can continue to iterate on it:

/approve
/lgtm

Before we do, let's make sure the remaining comment from Travis gets captured as a follow up we need to resolve before we move out of provisional.

/hold

robscott commented:
In the v0.8.0 API review session we got some pushback from SIG-Network API reviewers on the concept of this infrastructure stanza in the Gateway resource. I'd like to hold off on merging this until we can get some resolution within the broader SIG-Network community, likely at one of those meetings.

youngnick commented:
I think I'd summarize the API review feedback for this as `infrastructure` on Gateway trespasses a bit into the point of GatewayClass. The original intent of GatewayClass was for it to hold many of the type of settings that this GEP is talking about, and to use `paramsRef` for that purpose.

My initial feelings here are:
- Quite a few people would like to allow their users to create Gateways with slightly different properties without needing to enumerate all those properties in separate GatewayClasses - because GatewayClasses are only cluster scoped, granting access to them is higher-privilege than those people want to allow. So there is some demand for being able to control at least _some_ properties of Gateways at a non-GatewayClass level. (Particularly annotations and labels, I think).
- I think that we need to go further with the existing `paramsRef` and make _all_ use of `paramsRef` implementation-specific, including what happens if you combine GatewayClass `paramsRef` with Gateway `paramsRef`.
- We should instead allow only very specific fields with very explicitly defined behaviors in the `infrastructure` stanza, and have the same fields present on both GatewayClass and Gateway `infrastructure` stanzas, with very explicit rules about how the two sets of settings interact. If implementations want things outside that, they should use `paramsRef` and be implementation-specific.

Right now, it's _very_ common for in-cluster implementations to need to set annotations and/or labels on in-cluster generated resources, and we have no solution. The longer we let this situation continue, the more likely that we will end up with a defacto standard (like copying annotations and labels from the Gateway object to any generated objects), and need to recognize that here.

howardjohn commented:
> I think I'd summarize the API review feedback for this as `infrastructure` on Gateway trespasses a bit into the point of GatewayClass. The original intent of GatewayClass was for it to hold many of the type of settings that this GEP is talking about, and to use `paramsRef` for that purpose.

I *think* I addressed your comments in the latest commit. LMK

* ParamRef in Gateway, defined merging
* Annotations and Labels in GWC, defined merging

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+
+## Design
+
+This GEP does not introduce new API fields, but rather standardizes how implementations should behave when implementing the existing API.
```
This is not actually true anymore.

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+## Goals
+
+* Provide prescriptive guidance for how in-cluster implementations should behave.
+* Provide requirements (tested by conformance) for how in-cluster implementations should behave.
```
It seems like most if not all of this GEP would be pretty difficult to write conformance tests for, do you have any ideas for how this would work?

howardjohn replied:
That is a good point. Not really... any ideas?

robscott replied:
I think this is unfortunately going to be a feature that doesn't/can't have conformance tests.

robscott replied:
```suggestion
* Provide requirements for how in-cluster implementations should behave.
```

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+
+Gateway API provides a common abstraction over different implementations, whether they are implemented by cloud load balancers, in-cluster deployments, or other mechanisms. However, many in-cluster implementations have solved some of the same problems in different ways.
+
+Related discussions:
+* [Support cluster-local Gateways](https://github.com/kubernetes-sigs/gateway-api/discussions/1247)
```
New line require for list formatting to work:
```suggestion
Related discussions:

* [Support cluster-local Gateways](https://github.com/kubernetes-sigs/gateway-api/discussions/1247)
```

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+  This is *often*, but not necessarily, by deploying a `Deployment`/`DaemonSet` and `Service`.
+* Automated deployment: refers to an implementation that automatically deploys the data plane based on a `Gateway`.
+  That is, the user simply creates a `Gateway` resource and the rest is handled behind the scenes by the implementation.
+* Manual deployment: refers to an implementation that does NOT automatically deploy the data plane based on a `Gateway`.
+  This may require, for example, a user to manually create a `Deployment` and `Service` for the data plane, then link this to a `Gateway`.
```
I think we should just leave this out entirely. I don't think we've formally recognized this form of deployment in a GEP and I'd rather only do that if we had some clear guidelines for how it should be handled (or just say that this is a mode we don't want to support/encourage).

robscott commented on geps/gep-1762.md:
````diff
@@ -0,0 +1,294 @@
+    protocol: HTTP
+```
+
+With this configuration, an implementation:
+* MUST mark the Gateway as `Programmed` and provide an address in `Status.Addresses` where the Gateway can be reached on each configured port.
````
New line for list formatting:
```suggestion
With this configuration, an implementation:

* MUST mark the Gateway as `Programmed` and provide an address in `Status.Addresses` where the Gateway can be reached on each configured port.
```

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+
+With this configuration, an implementation:
+* MUST mark the Gateway as `Programmed` and provide an address in `Status.Addresses` where the Gateway can be reached on each configured port.
+* MUST label all generated resources (Service, Deployment, etc) with `gateway.networking.k8s.io/metadata.name: my-gateway` (where `my-gateway` is the name of the Gateway resource).
```
I think I'd prefer `gateway-name` here because `gateway.networking.k8s.io` refers to the entire Gateway API group, not just the Gateway resource.

```suggestion
* MUST label all generated resources (Service, Deployment, etc) with `gateway.networking.k8s.io/gateway-name: my-gateway` (where `my-gateway` is the name of the Gateway resource).
```

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+* MUST name all generated resources `my-gateway-example` (`<NAME>-<GATEWAY CLASS>`).
+  This is not simply `NAME` to reduce the chance of conflicts with existing resources.
+
+### Manual Deployments
+
+Managing "Manual Deployments" is currently out of scope for this GEP, but may be added in the future.
```
IMO we don't need this.

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+
+Managing "Manual Deployments" is currently out of scope for this GEP, but may be added in the future.
+
+### Customizations
```
I think ~everything in this section that isn't related to labels and annotations can probably be removed now.

howardjohn replied:
This is the background part, I think its useful to say what things folks may want to customize even if we don't solve all of them? I updated this to say they are not all addressed.

robscott replied:
That's fair, resolving this one.

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+This is typically not a concern; however, if an implementation is aware of specific meanings of certain labels or annotations, they MAY
+exclude these from irrelevant resources.
+
+This is intended to support integration with the kitchen-sync of Kubernetes extensions which rely on labels and annotations,
+such as as prometheus scraping, object grouping/organization, etc.
+
```
I think the least controversial use of this field is going to be to clearly identify resources that are associated with a specific application, environment, or Gateway. I'd recommend leading with that and then also mentioning this as an additional use case.

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+
+### Resource Attachment
+
+Resources Generated in response to the `Gateway` are guaranteed to have two attributes:
+* A `gateway.networking.k8s.io/metadata.name: <NAME>` label.
```
New line for list formatting
```suggestion
Resources Generated in response to the `Gateway` are guaranteed to have two attributes:

* A `gateway.networking.k8s.io/metadata.name: <NAME>` label.
```

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+
+Resources Generated in response to the `Gateway` are guaranteed to have two attributes:
+* A `gateway.networking.k8s.io/metadata.name: <NAME>` label.
+* A name `<NAME>-<GATEWAY CLASS>`.
```
Now I'm seeing the purpose of an exact name requirement, I don't think we can guarantee it though. Maybe we should say that implementations should use this as long as the name is not already claimed, and then note that here?

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+
+
+These can be relied on to attach resources to.
+While "Policy attachment" in the gateway-api would use attachment to the actual `Gateway` resource itself,
```
```suggestion
While "Policy attachment" in Gateway API would use attachment to the actual `Gateway` resource itself,
```

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+
+These can be relied on to attach resources to.
+While "Policy attachment" in the gateway-api would use attachment to the actual `Gateway` resource itself,
+many existing things attach only to resources like `Deployment` or `Service`.
```
```suggestion
many existing resources attach only to resources like `Deployment` or `Service`.
```

robscott commented on geps/gep-1762.md:
````diff
@@ -0,0 +1,294 @@
+      gateway.networking.k8s.io/metadata.name: gateway
+```
+
+Note: there is [discussion](https://github.com/kubernetes-sigs/gateway-api/discussions/1355) around a way to attach a HPA to a Gateway directly.
````
Mind adding this to a new section called "Future Work" at the bottom of the GEP? I'd like to make sure it's obvious that we don't want to be stuck with attaching an HPA to a Deployment forever.

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+	// For implementations creating other Kubernetes objects, this should be the `metadata.labels` field on resources.
+	// For other implementations, this refers to any relevant (implementation specific) "labels" concepts.
+	//
+  // If Labels is set on the GatewayClass as well, the labels are merged with the Gateway's labels taking precedence.
+  //
```
Can we just start with this on the Gateway and consider support on `GatewayClass` as future work? I think we could make a good argument for precedence to go in either direction here and/or to recreate the `overrides` and `defaults` concept from inherited policy attachment in the `GatewayClass.spec.infrastructure` field to enable both. I think that's a sufficiently complex conversation that it can be saved for a follow up and just removed from the initial scope of this GEP.

howardjohn replied:
I had it like that at first and @youngnick suggesting it would be a good requirement. I don't have a preference either way so will do whatever

robscott replied:
@youngnick's traveling for a bit so may take some time to respond, but I did get some time to talk with him about this yesterday and he was on board with only targeting Gateway to start.

youngnick replied:
Yes, this is on me, I think that having this work for Gateway first is better. Let's get something first, and add the GatewayClass later if needed.

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+	//
+  // If Labels is set on the GatewayClass as well, the labels are merged with the Gateway's labels taking precedence.
+  //
+	// Support: Implementation-specific
```
I think the patterns we're describing here would actually be "Extended" since we'd expect the implementations that support this to support it in the same way.
```suggestion
	// Support: Extended
```

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+	...
+}
+
+// LocalParametersReference identifies an API object containing controller-specific
+// configuration resource within the cluster.
+type LocalParametersReference struct {
+	// Group is the group of the referent.
+	Group Group `json:"group"`
+
+	// Kind is kind of the referent.
+	Kind Kind `json:"kind"`
+
+	// Name is the name of the referent.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Name string `json:"name"`
+}
```
I don't think this is needed anymore.

howardjohn replied:
Its in GatewayInfrastructure, so probably pending decision on https://github.com/kubernetes-sigs/gateway-api/pull/1757#discussion_r1327657519?

robscott replied:
Yeah, I'd personally rather keep this GEP as focused as possible since we're really only a couple of weeks away from trying to be feature complete for v1.0. I don't really have anything against ParamsRef here, but similar to the comment above, it introduces some form of merging logic between GWC and GW, and I want to make sure we take the time to get it right so would rather leave it out of scope for now.

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+	...
+}
+
+type GatewayClassInfrastructure struct {
```
Vote to remove this.

howardjohn replied:
Lets discuss in https://github.com/kubernetes-sigs/gateway-api/pull/1757#discussion_r1327657519, I think they are basically the same suggestion

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,294 @@
+	// For implementations creating other Kubernetes objects, this should be the `metadata.annotations` field on resources.
+	// For other implementations, this refers to any relevant (implementation specific) "annotations" concepts.
+	//
+  // If Annotations is set on the GatewayClass as well, the labels are merged with the Gateway's labels taking precedence.
+  //
```
Indentation is inconsistent in some of these examples, I'm assuming it's a mix of tabs and spaces?

robscott reviewed (commented):
Thanks @howardjohn! I think this will be a great addition to the API and should help provide more consistency for in-cluster implementations of the API. A lot of comments but mostly nits/very small things.

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,299 @@
+
+This follows the same behavior as out-of-cluster Gateway implementations.
+
+Note: this differs from the selection of "manual deployment", which doesn't use `IPAddress` type.
+Instead, these Gateways attach to some existing infrastructure (such as Service) which provides an address.
```
Nit: Would probably remove this as it refers to "manual deployment" which no longer exists in this GEP.

robscott commented on geps/gep-1762.md:
````diff
@@ -0,0 +1,299 @@
+    protocol: HTTP
+```
+
+This would generate a `Service` with `clusterIP` or `loadBalancerIP`, depending on the Service type.
````
I can't really tell if this is trying to tell implementations what they should do, or just provide an example of how some implementations handle this.

howardjohn replied:
The intent here got muddled through the revisions. Its meant to be "If they provide an IP, you better use it" and whether you use it as the clusterIP vs the loadBalancerIP is based on the type of service, which was determined by routability previously

robscott replied:
I think this is already covered reasonably well in the spec:

https://github.com/kubernetes-sigs/gateway-api/blob/cd223ea715cd08bcaac6f8524dd9302199bc362b/apis/v1beta1/gateway_types.go#L134-L163

It may be worth describing how in-cluster implementations can accomplish this in this GEP, but I don't think we need to add any new requirements here. Just trying to avoid introducing too many unique sources of requirements in case they need to change in the future.

howardjohn replied:
I made it more clear, I think. LMK what you think

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,299 @@
+
+This would generate a `Service` with `clusterIP` or `loadBalancerIP`, depending on the Service type.
+
+This follows the same behavior as out-of-cluster Gateway implementations.
```
I'm not really sure I see the parallel here? They may be providing similar functionality, but I don't think the way it's accomplished is very similar.

howardjohn replied:
is it more clear with https://github.com/kubernetes-sigs/gateway-api/pull/1757/files/4c2ae889c6ce055173e59aad70a3f31e06ccabff#r1327844348?

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,299 @@
+
+Resources Generated in response to the `Gateway` will have two attributes:
+
+* A `gateway.networking.k8s.io/gateway-name: <NAME>` label.
```
What happens if someone specifies this in `infrastructure.labels`? Maybe we should add validation to the labels and annotations infra fields that prevents the use of any keys that include `gateway.networking.k8s.io`?

howardjohn replied:
:+1: , will add that

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,299 @@
+	// If Labels is set on the Gateway as well, the labels are merged with the Gateway's labels taking precedence.
+	//
+	// Support: Extended
+	Labels map[string]string `json:"labels,omitempty"`
```
Assuming we'll copy validation from upstream here, don't need to figure that out in this GEP, but it might be useful to at least describe the max number of entries in each map. I'd prefer to start with something small like 8 unless we have a clear use case for more.

howardjohn replied:
I put 16 max

robscott replied:
Any specific use cases that would benefit from more than 8?

howardjohn replied:
I just vaguely recall seeing some pretty monstrous set of annotations for AWS/Azure and it seemed pretty minimal risk to make it 16 tbh. So no strong case really.

robscott replied:
There definitely are some massive lists out there (for example https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.2/guide/service/annotations/), but I don't think it's common to configure all of these at the same time. I'd personally rather start with a smaller number here and only increase if/when needed. We've left room in the API to loosen validation in future releases, but not tighten it. 

I know it probably has limited efficacy, but somehow it feels like the tighter our validation is here, the less likely this field will end up being overloaded for unintended/unexpected purposes.

robscott reviewed (commented):
Thanks for the quick updates @howardjohn! Just a few more nits.

howardjohn commented:
@robscott GatewayClass dropped

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,245 @@
+This is typically not a concern; however, if an implementation is aware of specific meanings of certain labels or annotations, they MAY
+exclude these from irrelevant resources.
+
+This is intended to clearly identify resources associatd with a specific application, environment, or Gateway.
```
```suggestion
This is intended to clearly identify resources associated with a specific application, environment, or Gateway.
```

robscott commented on geps/gep-1762.md:
````diff
@@ -0,0 +1,245 @@
+
+```go
+type GatewayInfrastructure struct {
+	// Labels that should be applied to any resources created in response to this Gateway.
````
One thing that is not clear to me is if this is intended to be an exclusive set or if the implementation can choose to add additional labels or annotations. I'm assuming the latter, but should make it clear in the spec.

robscott commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,245 @@
+
+The generated resources MUST be in the same namespaces as the `Gateway`.
+
+Implementations MAY set `ownerReferences` to the `Gateway` in most cases, as well, but this is not required
+as some implementations may have different cleanup mechanisms.
+
+
+These can be relied on to attach resources to.
+While "Policy attachment" in Gateway API would use attachment to the actual `Gateway` resource itself,
```
Nit: The "These can be relied on to attach resources to" sentence feels a bit out of place here. I'm assuming you're not referring to the owner refs described in the sentence above, but I'm not sure what "these" means in this context. Also the extra new line above and lacking one below feels a bit wonky.

robscott reviewed (commented):
Thanks @howardjohn!

shaneutt reviewed (approved):
/approve
/lgtm

arkodg commented on geps/gep-1762.md:
```diff
@@ -0,0 +1,248 @@
+This section just clarifies and existing part of the spec, how to handle `.spec.addresses` for in-cluster implementations.
+Like all other Gateway types, this should impact the address the `Gateway` is reachable at.
+
+For implementations using a `Service`, this means the `clusterIP` or `loadBalancerIP` (depending on the `Service` type).
```
`ExternalIP` may also be valid here for `NodePort` case

robscott replied:
Agreed. Don't want to block this merging but would be great to cover this in a follow up.

robscott commented:
Thanks @howardjohn!

/hold cancel
/lgtm
